Sprint 9 solution

Naz1anmak opened the pull request:
- Настроил доступ к методам программы через HTTP API
- Написаны дополнительные классы тестов

Naz1anmak committed:
Version 6.0
- Настроил доступ к методам программы через HTTP API
- Написаны дополнительные классы тестов

Naz1anmak committed:
Version 6.1
- Теперь вместе null в TaskManager пробрасывается NotFoundException
- в handlers добавлены дополнительные блоки try-catch
- Переписаны тесты

afomkina commented on src/manager/InMemoryTaskManager.java:
```diff
@@ -333,14 +337,14 @@ public void updateSubtask(Subtask newSubtask) {
     @Override
     public void deleteSubtaskById(int id) {
-        Subtask subtask = subtasks.remove(id);
+        Subtask subtask = subtasks.get(id);
```
Нужно вернуть `Subtask subtask = subtasks.remove(id);`. Если задача не будет найдена, то как и в случае с get, вернется null. Соответственно, он работает точно также как и get, только при этом удаляет найденную задачу.
Поправь также для задач и эпиков

afomkina commented on src/server/adapters/Adapters.java:
```diff
@@ -0,0 +1,47 @@
+
+public class Adapters {
+    public static class LocalDateTimeAdapter extends TypeAdapter<LocalDateTime> {
+        private static final DateTimeFormatter dtf = DateTimeFormatter.ofPattern("dd.MM.yyyy HH:mm");
```
Все константы в java принято именовать в SNAKE_CASE. Лучше не использовать сокращения, так как они делают код менее читаемым

afomkina commented on src/server/HttpTaskServer.java:
```diff
@@ -0,0 +1,73 @@
+
+public class HttpTaskServer {
+    private static final int PORT = 8080;
+    private static TaskManager taskManager = null;
```
`static` у полей класса нужны только в случае, если мы хотим сделать значение константный (`static final`) или же нам нужен к ним доступ из статического контекста. В любом другом случае лучше избегать `static` у полей класса, так как такой код не соответствует парадигме ООП

Naz1anmak replied:
если убрать `static` у поля PORT, то придется менять само название. Потому что автотест на гитхабе ругается, что поле написано в SNAKE_CASE, но не является константой. Или менять название на `post` или оставлять `static`

afomkina commented on src/server/HttpTaskServer.java:
```diff
@@ -0,0 +1,73 @@
+public class HttpTaskServer {
+    private static final int PORT = 8080;
+    private static TaskManager taskManager = null;
+    private static HttpServer httpServer;
```
Нужно убрать модификатор `static`

afomkina commented on src/server/HttpTaskServer.java:
```diff
@@ -0,0 +1,73 @@
+        HttpTaskServer.taskManager = taskManager;
+    }
+
+    public static void main(String[] args) throws IOException {
```
Код запуска должен находиться в классе Main

afomkina commented on src/server/HttpTaskServer.java:
```diff
@@ -0,0 +1,73 @@
+        System.out.println("HTTP-сервер запущен на " + PORT + " порту!");
+    }
+
+    public void stop() {
```
Этот метод нужно использовать, чтобы остановить сервер

Naz1anmak replied:
Не совсем понял комментарий

afomkina commented on src/server/handlers/BaseHttpHandler.java:
```diff
@@ -0,0 +1,31 @@
+    }
+
+    protected void sendNotFound(HttpExchange exchange) throws IOException {
+        sendText(exchange, "Ошибка: путь не найден", 404);
```
Задание со `"*"`:
Можно создать отдельное перечисление для HTTP кодов и использовать его вместо "голых" цифр. 
Так код будет выглядеть более читаемо и понятней. 
Примеры можно посмотреть в главе [Конструкторы, поля и методы перечисления](https://metanit.com/java/tutorial/3.8.php#:~:text=%D0%9A%D0%BE%D0%BD%D1%81%D1%82%D1%80%D1%83%D0%BA%D1%82%D0%BE%D1%80%D1%8B%2C%20%D0%BF%D0%BE%D0%BB%D1%8F%20%D0%B8%20%D0%BC%D0%B5%D1%82%D0%BE%D0%B4%D1%8B%20%D0%BF%D0%B5%D1%80%D0%B5%D1%87%D0%B8%D1%81%D0%BB%D0%B5%D0%BD%D0%B8%D1%8F)

afomkina commented on src/server/handlers/HistoryHandler.java:
```diff
@@ -0,0 +1,51 @@
+
+    @Override
+    public void handle(HttpExchange exchange) throws IOException {
+        String method = exchange.getRequestMethod();
```
Код со строк 26-35 повторяется в классе `PrioritizedHandler`. Нужно избавиться от дублирований

afomkina commented on src/server/handlers/HistoryHandler.java:
```diff
@@ -0,0 +1,51 @@
+                .create();
+
+        List<Task> history = taskManager.getHistory();
+        switch (endpoint) {
```
Код со строк 38-49 повторяется в классе `PrioritizedHandler`. Нужно избавиться от дублирований

afomkina commented on src/server/handlers/TaskHandler.java:
```diff
@@ -0,0 +1,146 @@
+
+                } catch (NotFoundException e) {
+                    sendText(exchange, e.getMessage(), 404);
+                } catch (IllegalArgumentException e) {
```
Нужно избавиться от однобуквенных переменных в коде

afomkina reviewed (changes requested):
Клево, что написал так много тестов 👍 Молодец

Naz1anmak committed:
Version 6.2
- Запуск перенесен в класс Main
- Объединил HistoryHandler и PrioritizedHandler
   - Объединил тесты этих классов
- Создал enum HttpStatusCode для перечисления HTTP кодов
   - Переписаны handlers
- исправлены прочие ошибки
